datacube/utils/documents.py

```diff
@@ -506,42 +483,35 @@ def __setattr__(self, name, val):
         if offset is None:
             raise AttributeError(
                 'Unknown field offset %r. Expected one of %r' % (
-                    name, list(self._fields.keys())
+                    name, list(self.system_fields.keys())
                 )
             )
-        return _set_doc_offset(offset, self._doc, val)
+        self.doc = toolz.update_in(self.doc, offset, lambda _: val)
+
+    def __dir__(self):
+        return list(self.fields)
 
     @property
-    def fields(self):
-        fields = {}
-        fields.update(self.search_fields)
-        fields.update(self.system_fields)
-        return fields
+    def doc(self):
+        return self._doc
 
     @property
     def search_fields(self):
-        fields = {}
-        for name, field in self._search_fields.items():
-            try:
-                fields[name] = field.extract(self._doc)
-            except (AttributeError, KeyError, ValueError):
-                continue
-        return fields
+        return {name: field.extract(self.doc)
```

Comment:
Are there implications for not catching those errors?  It feels like a risky thing to move, but it seems things would have to be pretty broken elsewhere for any of those errors to occur.

Reply:
There's have to be a mismatch between the MetadataType definition and the Dataset's metadata document. If that doesn't already cause other problems elsewhere, it might be better to throw an exception or at least log a warning instead of letting it silently slip past.